### ISSUE TITLE
How to query Datasets registered in Manually created SAS Libraries

### ISSUE THREAD

**Issue by sudipto84goswami** [sassoftware/saspy#17]
Hi Team,

My requirement is to access data from a specific SAS Library.

Example: hr = sas.sasdata('air', 'sashelp') This statement works fine as we are querying data from sashelp which is a library internally created with SAS.

I need to access data from a library say XYZ and table ABC. So if I use 
hr1 = sas.sasdata('ABC', 'XYZ') this throws error as Table XYZ.ABC does not exist. But I can very well query the same from Enterprise Guide.

How do i replicate the SAS Query: 

proc contents data=XYZ.ABC;
run;

Please let me know if any further information is required.

**Comment by jld23** [comment]
@sudipto84goswami 
You can use the `saslib` method on the session to assign libraries
Here is link in the doc: https://sassoftware.github.io/saspy/api.html#saspy.SASsession.saslib

Here is an example assigning library xyz
`sas.saslib(libref = 'XYZ', path = '.')`
and here is the coresponding output
```
28   libname XYZ    '.'  ;
NOTE: Libref XYZ was successfully assigned as follows: 
      Engine:        V9 
      Physical Name: /opt/notebook
```

You can see from the doc there is also support for libnames to databases through SAS/ACCESS.

**Comment by sudipto84goswami** [comment]
Thanks Jared. How do I query after the library is being assigned. And can't we directly refer to Library which is already defined at SAS Server.

proc contents data=XYZ.ABC;
run;

**Comment by tomweber-sas** [comment]
Just use the sasdata method of the SASsession object to assign it and all the methods are available to you.

Sorry, I can't see the original from my phone right now, bit it would be something like this:

sas = ssspy.SASsession()

mydata = sas.sasdata('dataset', libref='libref')

mydata.contents()

If you hit Tab after sasdata.  (A sasdata object followed by for (period)) it will show you all the methods available. Contents is one of them.



▶ Show quoted text

**Comment by tomweber-sas** [comment]
Ok, I'm at a terminal now. I think you already have a libref pre-assigned on your SAS server; libref xyz that refers to your data. If not, use the saslib() method Jared showed to assign the libref. Then, use the sasdata method I showed, to get a python SASdata object that refers to that SAS Data Set (mydata =sas.sasdata('abc', libref='xyz')). That SASdata object then has lots of methods to interact with that SAS Data Set. SASdata.contents() is one of many. If there's something to add, we are open to that for sure. We haven't written every method that there ca be - yet. 

And, just in case: I recently (since the last Pypi build) pushed a fix for the contents method, when the results='Pandas' is in effect (which is the default now - I don't care for that, but that's what it is), so if you do a SASdata.contents() and it throws an errror, either set "results='HTML'"  in your SASsession(...) command, or pull the latest code for a fix, if you like pandas output instead of ODS HTML.

**Comment by tomweber-sas** [comment]
?Just use the sasdata method of the SASsession object to assign it and all the methods are available to you.

Sorry, I can't see the original from my phone right now, bit it would be something like this:

sas = ssspy.SASsession()

mydata = sas.sasdata('dataset', libref='libref')

mydata.contents()

If you hit Tab after sasdata.  (A sasdata object followed by for (period)) it will show you all the methods available. Contents is one of them.



On Mar 30, 2017 20:56, sudipto84goswami <notifications@github.com> wrote:

Thanks Jared. How do I query after the library is being assigned. And can't we directly refer to Library which is already defined at SAS Server.

proc contents data=XYZ.ABC;
run;

-

**Comment by tomweber-sas** [comment]
Following up on this:
the saslib() method did assign a library. the sasdata() method did return a SASdata object for the member. The contents() method will show the proc contents output with results='html', but there was a bug in it when results='Pandas' - which is fixed in the repo (40c508c8eeaa95b0793dfa27215d57ffdc2f7fcd) and confirmed at their site after updating with the fix.
The expected pre-assigned librefs were not, in fact, assigned, so we are looking into where this pre-assignment was being done so we can get it done for the saspy connections too.
Closing this issue as it is all working and we'll get the pre-assigned libraries figured out separately. It may be they are being assigned by EG at start-up, not in the workspace servers. Either way, once that's determined at the site, we can get them assigned for the saspy workspace servers too.

**Comment by sudipto84goswami** [comment]
Thanks a lot Tom.
